File: macos/Tests/ElevateCoreTests/PropagationWatcherTests.swift
import Foundation
import Testing
@testable import ElevateCore

@Suite struct PropagationWatcherTests {
    private let me = Identity(id: "id1", upn: "u@contoso.com", displayName: "U", homeTenantId: "t-home")
    private static let entraScope = RoleScope.entraDirectory(roleDefinitionId: "role-1", directoryScopeId: "/")
    private static let groupScope = RoleScope.group(groupId: "grp-1", accessId: .member)

    private func active(_ scope: RoleScope) -> ActiveAssignment {
        ActiveAssignment(roleKey: RoleKey(identityId: "id1", tenantId: "t1", scope: scope),
                         assignmentId: "a1", startDateTime: .now, endDateTime: nil, status: .active)
    }

    /// A provider whose probe answers from a queue, so a test can make propagation take a few rounds.
    actor ProbeState {
        private var answers: [EffectiveAccess]
        private(set) var probes = 0
        private var failure: PIMError?

        init(_ answers: [EffectiveAccess]) { self.answers = answers }
        func setFailure(_ e: PIMError?) { failure = e }
        func next() throws -> EffectiveAccess {
            probes += 1
            if let failure { throw failure }
            return answers.isEmpty ? .notYet : answers.removeFirst()
        }
    }

    struct ProbeProvider: PIMProvider {
        let kind: RoleScopeKind
        let scopes: [String] = []
        let state: ProbeState

        init(kind: RoleScopeKind, answers: [EffectiveAccess] = []) {
            self.kind = kind
            state = ProbeState(answers)
        }

        func effectiveAccess(_ assignment: ActiveAssignment, identity: Identity) async throws -> EffectiveAccess {
            try await state.next()
        }

        func eligibleRoles(identity: Identity, tenant: TenantContext) async throws -> [EligibleRole] { [] }
        func activeAssignments(identity: Identity, tenant: TenantContext) async throws -> [ActiveAssignment] { [] }
        func policy(for role: EligibleRole, identity: Identity) async throws -> RolePolicy { .manualDefault }
        func activate(_ request: ActivationRequest, identity: Identity) async throws -> ActiveAssignment {
            throw PIMError.notEligible
        }
        func deactivate(_ assignment: ActiveAssignment, identity: Identity) async throws {}
        func cancelPendingRequest(_ assignment: ActiveAssignment, identity: Identity) async throws {}
    }

    /// Records the delays instead of sleeping, so the tests run in no time.
    actor Clock {
        private(set) var elapsed: TimeInterval = 0
        func advance(_ seconds: TimeInterval) { elapsed += seconds }
    }

    private func watcher(_ clock: Clock, _ providers: [any PIMProvider]) -> PropagationWatcher {
        PropagationWatcher(
            coordinator: ActivationCoordinator(providers: providers, tokens: FakeTokenProvider()),
            sleep: { await clock.advance($0) })
    }

    @Test func readyOnceTheProbeConfirms() async throws {
        let clock = Clock()
        let provider = ProbeProvider(kind: .entraDirectory, answers: [.notYet, .confirmed])

        let outcomes = try await watcher(clock, [provider]).wait(for: [active(Self.entraScope)], identities: [me])

        #expect(outcomes.count == 1)
        #expect(outcomes.first?.state == .ready)
        #expect(await provider.state.probes == 2)
    }

    @Test func reportsEachRoleAsItSettlesRatherThanOnlyAtTheEnd() async throws {
        let clock = Clock()
        let reported = Reported()
        let provider = ProbeProvider(kind: .entraDirectory, answers: [.confirmed])

        _ = try await watcher(clock, [provider])
            .wait(for: [active(Self.entraScope)], identities: [me], onChange: { outcome in reported.add(outcome) })

        #expect(reported.all.count == 1)
        #expect(reported.all.first?.state == .ready)
    }

    /// `onChange` is not async, so the collector guards itself.
    final class Reported: @unchecked Sendable {
        private let lock = NSLock()
        private var items: [PropagationOutcome] = []
        func add(_ o: PropagationOutcome) { lock.lock(); items.append(o); lock.unlock() }
        var all: [PropagationOutcome] { lock.lock(); defer { lock.unlock() }; return items }
    }

    @Test func unconfirmedOnceTheDeadlinePasses() async throws {
        let clock = Clock()
        let provider = ProbeProvider(kind: .entraDirectory)

        let outcomes = try await watcher(clock, [provider])
            .wait(for: [active(Self.entraScope)], identities: [me], deadline: 0)

        #expect(outcomes.count == 1)
        #expect(outcomes.first?.state == .unconfirmed)
        // The user is told what usually explains it, not just that it did not happen.
        #expect(outcomes.first?.detail == PropagationHints.likelyCause(.entraDirectory))
    }

    @Test func unobservableStopsProbingStraightAway() async throws {
        let clock = Clock()
        let provider = ProbeProvider(kind: .entraDirectory, answers: [.unknown("scoped role")])

        let outcomes = try await watcher(clock, [provider]).wait(for: [active(Self.entraScope)], identities: [me])

        #expect(outcomes.first?.state == .unobservable)
        #expect(await provider.state.probes == 1)
    }

    @Test func aFailedProbeIsUnobservableNotAFailedActivation() async throws {
        let clock = Clock()
        let provider = ProbeProvider(kind: .entraDirectory)
        await provider.state.setFailure(.network("offline"))

        let outcomes = try await watcher(clock, [provider]).wait(for: [active(Self.entraScope)], identities: [me])

        #expect(outcomes.first?.state == .unobservable)
    }

    @Test func oneUnobservableRoleDoesNotHoldUpTheOthers() async throws {
        let clock = Clock()
        let entra = ProbeProvider(kind: .entraDirectory, answers: [.unknown("scoped role")])
        let group = ProbeProvider(kind: .group, answers: [.notYet, .confirmed])

        let outcomes = try await watcher(clock, [entra, group])
            .wait(for: [active(Self.entraScope), active(Self.groupScope)], identities: [me])

        #expect(outcomes.count == 2)
        #expect(outcomes.first { $0.roleKey.scope.kind == .entraDirectory }?.state == .unobservable)
        #expect(outcomes.first { $0.roleKey.scope.kind == .group }?.state == .ready)
        // The Entra role was asked once and then left alone.
        #expect(await entra.state.probes == 1)
    }

    @Test func onlyActiveAssignmentsAreProbed() async throws {
        let clock = Clock()
        let provider = ProbeProvider(kind: .entraDirectory, answers: [.confirmed])
        let pending = ActiveAssignment(
            roleKey: RoleKey(identityId: "id1", tenantId: "t1", scope: Self.entraScope),
            assignmentId: "a1", startDateTime: .now, endDateTime: nil, status: .pendingApproval)

        let outcomes = try await watcher(clock, [provider]).wait(for: [pending], identities: [me])

        #expect(outcomes.isEmpty)
        #expect(await provider.state.probes == 0)
    }

    @Test func anAccountThatIsNotSignedInHereIsUnobservable() async throws {
        let clock = Clock()
        let provider = ProbeProvider(kind: .entraDirectory, answers: [.confirmed])

        let outcomes = try await watcher(clock, [provider]).wait(for: [active(Self.entraScope)], identities: [])

        #expect(outcomes.first?.state == .unobservable)
        #expect(await provider.state.probes == 0)
    }

    @Test func aRoleWithNoProviderIsUnobservable() async throws {
        let clock = Clock()
        let outcomes = try await watcher(clock, [ProbeProvider(kind: .group, answers: [.confirmed])])
            .wait(for: [active(Self.entraScope)], identities: [me])

        #expect(outcomes.first?.state == .unobservable)
    }

    @Test func probingWaitsBeforeTheFirstCallAndThenSettlesOnTheLongerInterval() async throws {
        let clock = Clock()
        let provider = ProbeProvider(kind: .entraDirectory, answers: [.notYet, .confirmed])

        _ = try await watcher(clock, [provider]).wait(for: [active(Self.entraScope)], identities: [me])

        #expect(await clock.elapsed == PropagationWatcher.defaultFirstInterval + PropagationWatcher.defaultMaxInterval)
    }
}

File: windows/tests/Elevate.Core.Tests/PropagationWatcherTests.cs
using Elevate.Core.Coordination;
using Elevate.Core.Models;
using Elevate.Core.Providers;
using Elevate.Core.Tests.Support;
using FluentAssertions;

namespace Elevate.Core.Tests;

/// <summary>Port of the Swift <c>PropagationWatcherTests</c>.</summary>
public class PropagationWatcherTests
{
    private static readonly Identity Me = new("id1", "u@contoso.com", "U", "t-home");

    private static ActiveAssignment Active(RoleScope scope) =>
        new(new RoleKey("id1", "t1", scope), "a1", DateTimeOffset.UtcNow, null, AssignmentStatus.Active);

    private static readonly RoleScope EntraScope = new EntraDirectoryScope("role-1", "/");
    private static readonly RoleScope GroupScopeValue = new GroupScope("grp-1", GroupAccess.Member);

    /// <summary>A provider whose probe answers from a queue, so a test can make propagation take a few rounds.</summary>
    private sealed class ProbeProvider(RoleScopeKind kind, params EffectiveAccess[] answers) : IPimProvider
    {
        private readonly Queue<EffectiveAccess> _answers = new(answers);

        public int Probes { get; private set; }

        public RoleScopeKind Kind => kind;

        public IReadOnlyList<string> Scopes => [];

        public Exception? Throws { get; set; }

        public Task<EffectiveAccess> EffectiveAccessAsync(ActiveAssignment assignment, Identity identity, CancellationToken ct = default)
        {
            Probes++;
            if (Throws is { } error)
            {
                return Task.FromException<EffectiveAccess>(error);
            }

            return Task.FromResult(_answers.Count > 0 ? _answers.Dequeue() : EffectiveAccess.NotYet);
        }

        public Task<IReadOnlyList<EligibleRole>> EligibleRolesAsync(Identity identity, TenantContext tenant, CancellationToken ct = default)
            => Task.FromResult<IReadOnlyList<EligibleRole>>([]);

        public Task<IReadOnlyList<ActiveAssignment>> ActiveAssignmentsAsync(Identity identity, TenantContext tenant, CancellationToken ct = default)
            => Task.FromResult<IReadOnlyList<ActiveAssignment>>([]);

        public Task<RolePolicy> PolicyAsync(EligibleRole role, Identity identity, CancellationToken ct = default)
            => Task.FromResult(RolePolicy.ManualDefault);

        public Task<ActiveAssignment> ActivateAsync(ActivationRequest request, Identity identity, CancellationToken ct = default)
            => Task.FromException<ActiveAssignment>(new PimException(PimErrorKind.NotEligible));

        public Task DeactivateAsync(ActiveAssignment assignment, Identity identity, CancellationToken ct = default) => Task.CompletedTask;

        public Task CancelPendingRequestAsync(ActiveAssignment assignment, Identity identity, CancellationToken ct = default) => Task.CompletedTask;
    }

    /// <summary>Advances a virtual clock instead of sleeping, so the tests run in no time.</summary>
    private sealed class Clock
    {
        public TimeSpan Elapsed { get; private set; }

        public Task DelayAsync(TimeSpan span, CancellationToken ct)
        {
            Elapsed += span;
            return Task.CompletedTask;
        }
    }

    private static PropagationWatcher Watcher(Clock clock, params IPimProvider[] providers) =>
        new(new ActivationCoordinator(providers, new FakeTokenProvider()), clock.DelayAsync);

    [Fact]
    public async Task ReadyOnceTheProbeConfirms()
    {
        var clock = new Clock();
        var provider = new ProbeProvider(RoleScopeKind.EntraDirectory, EffectiveAccess.NotYet, EffectiveAccess.Confirmed);

        var outcomes = await Watcher(clock, provider).WaitAsync([Active(EntraScope)], [Me]);

        outcomes.Should().ContainSingle().Which.State.Should().Be(PropagationState.Ready);
        provider.Probes.Should().Be(2);
    }

    [Fact]
    public async Task ReportsEachRoleAsItSettlesRatherThanOnlyAtTheEnd()
    {
        var clock = new Clock();
        var reported = new List<PropagationOutcome>();
        var provider = new ProbeProvider(RoleScopeKind.EntraDirectory, EffectiveAccess.Confirmed);

        await Watcher(clock, provider).WaitAsync([Active(EntraScope)], [Me], null, reported.Add);

        reported.Should().ContainSingle().Which.State.Should().Be(PropagationState.Ready);
    }

    [Fact]
    public async Task UnconfirmedOnceTheDeadlinePasses()
    {
        var clock = new Clock();
        var provider = new ProbeProvider(RoleScopeKind.EntraDirectory);

        var outcomes = await Watcher(clock, provider)
            .WaitAsync([Active(EntraScope)], [Me], TimeSpan.Zero);

        var outcome = outcomes.Should().ContainSingle().Subject;
        outcome.State.Should().Be(PropagationState.Unconfirmed);
        // The user is told what usually explains it, not just that it did not happen.
        outcome.Detail.Should().Be(PropagationHints.LikelyCause(RoleScopeKind.EntraDirectory));
    }

    [Fact]
    public async Task UnobservableStopsProbingStraightAway()
    {
        var clock = new Clock();
        var provider = new ProbeProvider(RoleScopeKind.EntraDirectory, EffectiveAccess.Unknown("scoped role"));

        var outcomes = await Watcher(clock, provider).WaitAsync([Active(EntraScope)], [Me]);

        outcomes.Should().ContainSingle().Which.State.Should().Be(PropagationState.Unobservable);
        provider.Probes.Should().Be(1);
    }

    [Fact]
    public async Task AFailedProbeIsUnobservable_NotAFailedActivation()
    {
        var clock = new Clock();
        var provider = new ProbeProvider(RoleScopeKind.EntraDirectory) { Throws = new PimException(PimErrorKind.Network, "offline") };

        var outcomes = await Watcher(clock, provider).WaitAsync([Active(EntraScope)], [Me]);

        outcomes.Should().ContainSingle().Which.State.Should().Be(PropagationState.Unobservable);
    }

    [Fact]
    public async Task OneUnobservableRoleDoesNotHoldUpTheOthers()
    {
        var clock = new Clock();
        var entra = new ProbeProvider(RoleScopeKind.EntraDirectory, EffectiveAccess.Unknown("scoped role"));
        var group = new ProbeProvider(RoleScopeKind.Group, EffectiveAccess.NotYet, EffectiveAccess.Confirmed);

        var outcomes = await Watcher(clock, entra, group)
            .WaitAsync([Active(EntraScope), Active(GroupScopeValue)], [Me]);

        outcomes.Should().HaveCount(2);
        outcomes.Single(o => o.RoleKey.Scope.Kind == RoleScopeKind.EntraDirectory).State
            .Should().Be(PropagationState.Unobservable);
        outcomes.Single(o => o.RoleKey.Scope.Kind == RoleScopeKind.Group).State
            .Should().Be(PropagationState.Ready);
        // The Entra role was asked once and then left alone.
        entra.Probes.Should().Be(1);
    }

    [Fact]
    public async Task OnlyActiveAssignmentsAreProbed()
    {
        var clock = new Clock();
        var provider = new ProbeProvider(RoleScopeKind.EntraDirectory, EffectiveAccess.Confirmed);
        var pending = new ActiveAssignment(
            new RoleKey("id1", "t1", EntraScope), "a1", DateTimeOffset.UtcNow, null, AssignmentStatus.PendingApproval);

        var outcomes = await Watcher(clock, provider).WaitAsync([pending], [Me]);

        outcomes.Should().BeEmpty();
        provider.Probes.Should().Be(0);
    }

    [Fact]
    public async Task AnAccountThatIsNotSignedInHereIsUnobservable()
    {
        var clock = new Clock();
        var provider = new ProbeProvider(RoleScopeKind.EntraDirectory, EffectiveAccess.Confirmed);

        var outcomes = await Watcher(clock, provider).WaitAsync([Active(EntraScope)], []);

        outcomes.Should().ContainSingle().Which.State.Should().Be(PropagationState.Unobservable);
        provider.Probes.Should().Be(0);
    }

    [Fact]
    public async Task ARoleWithNoProviderIsUnobservable()
    {
        var clock = new Clock();
        var outcomes = await Watcher(clock, new ProbeProvider(RoleScopeKind.Group, EffectiveAccess.Confirmed))
            .WaitAsync([Active(EntraScope)], [Me]);

        outcomes.Should().ContainSingle().Which.State.Should().Be(PropagationState.Unobservable);
    }

    [Fact]
    public async Task ProbingWaitsBeforeTheFirstCallAndThenSettlesOnTheLongerInterval()
    {
        var clock = new Clock();
        var provider = new ProbeProvider(RoleScopeKind.EntraDirectory, EffectiveAccess.NotYet, EffectiveAccess.Confirmed);
        var watcher = Watcher(clock, provider);

        await watcher.WaitAsync([Active(EntraScope)], [Me]);

        clock.Elapsed.Should().Be(watcher.FirstInterval + watcher.MaxInterval);
    }
}
